### PR TITLE
[READY]Gas Miners

### PR DESCRIPTION
**Pull request by silicons**
Ported from /VG/station.
**THIS PR WILL NOT ADD THEM TO THE MAP**

Vents that spew out gas and consume power from either an APC or a power node under it to function.
Intended to replace the super high pressure gas with roundstart low pressure but powerful gas miners to fill up the tanks and keep them filled later, but this does not currently change the game and only adds the gas miners.
Images "coming soon"(tm).

BASICALLY: ATMOSPHERICS WILL NEVER RUN OUT OF GAS UNLESS THEY ARE OUT OF POWER (BUT WILL RUN OUT MODERATELY FAST IF THEY ARE), ONCE THIS IS PUT ON THE MAP. MORE DEPENDENCY ON POWER. Even if this isn't put on the map/used to make power, this is probably better than "Oh this guy broke a SINGLE TILE ON THE TANKS! ROUNDS OVER, NO MORE OXYGEN, GO HOME!".

### CODE REVIEW

**Comment by Ninja137** [comment]
Alright, so where does the gas actually get mined from, though? Power just magically generates it from nowhere?

**Comment by kevinz000** [comment]
@Ninja137 The space station is in orbit around a planet.
I think.
RIGHT GUYS?!!
QUICK SPEW BULLSHIT!!
No really the official fluff from vg that would probably apply to us too is that we're in orbit so it's taking it from the planet.

**Comment by jughu** [comment]
Can you wrench gas miners like in vg and put them anywhere you like?

**Comment by kevinz000** [comment]
@jughu Possibly in the future but if this is added I really don't want people to be able to drag a plasma miner and an oxygen miner into the central hallways, set up a walled in pacman with tons of fuel, wire it up with reinforced floors, and laugh as something rivalling plasma fusion burns the station down :^)

**Comment by kevinz000** [comment]
This only adds the thing to the game, not to the maps. we can discuss mapping later, and if you want them to be possibly BUILDABLE, and MOVABLE, someone's going to have to code a REMOTE C ONTROLLED INTERFACE for them so they're not completely unstoppable once wrenched down and powered, as this can get out of hand really quick. Also, the power stuff and other requirements will have to be possibly balanced some more, but for now consider this a possible mapping object to replace all of the station's air coming from 1000000kpa tanks.

**Comment by kevinz000** [comment]
i'd code an interface and control system in this PR but I still can't even into TGUI (Holy shit even robustin's learned it shame on me) so this is still only a mapping thing, maybe when planetstation comes around I'll have the drive to learn TGUI (or if I need to code another feature that REQUIRES an interface no matter what.)

**Comment by jughu** [comment]
Someday atmos autism rule all (also make thermo generator orderable when?)

**Comment by kevinz000** [comment]
@jughu when someone makes its code not shit (i could but im working on tons more stuff too busy rip)

**Comment by Cebasiz** [comment]
it's a good idea

these things should be buildable btw, so you can construct more if they are destroyed or create more to serve a demand.

ideally atmos should have to expand a bit to fuel high-energy supermatter setups.

also when are we going to get the ability to create solid plasma for science with a high-energy reaction? thx

File: code/modules/atmospherics/machinery/other/miner.dm [hunk 1]
```diff
@@ -0,0 +1,152 @@
+/obj/machinery/atmospherics/miner
+	name = "gas miner"
+	desc = "Gasses mined from the gas giant below (above?) flow out through this massive vent."
```
Comment (GunHog) [inline code/modules/atmospherics/machinery/other/miner.dm]: But we have Lavaland planet now. And soon-possibly-maybe the station will be ON the planet!

Comment (Supermichael777) [inline code/modules/atmospherics/machinery/other/miner.dm]: Then we can change these to hydraulic fracturing rigs

Comment (kevinz000) [inline code/modules/atmospherics/machinery/other/miner.dm]: Ehhhhh yah it needs better fluff text

**Comment by FantasticFwoosh** [comment]
Why isn't this just a purposeful industrial grade chem fabricator? You're literally ripping off VG's lore of harvesting gas giants (which there are non in sight) for no good reason.

> We have borderline magic power = chemical converters that are star trek tier technology for printing off periodic table elements (in some form) but you want to 'mine' stuff that doesn't exist. Wat.

**Comment by YotaXP** [comment]
I'd say gasses should be obtainable from the asteroid, but I'm not sure how it would be transported back.  Normal canisters wouldn't cut it.

**Comment by FantasticFwoosh** [comment]
Asteroid doesn't exist anymore, so moon rock is out of the window.

File: code/modules/atmospherics/machinery/other/miner.dm (hunk 1 above)
Comment (Supermichael777) [inline code/modules/atmospherics/machinery/other/miner.dm]: magnetic gas net projector. https://en.wikipedia.org/wiki/Bussard_ramjet

**Comment by Iamgoofball** [comment]
who gives a fuck about lore lmao

this is good :+1:

**Comment by Anonmare** [comment]
Obviously they acquire gas through high energy reactons known as Spontaneous Bluespace Quantum Molecularisation. 
As we all know, matter and energy are intrinsically the same thing and that you cannot create matter from nothing. This reaction uses the strange qualities of bluespace to greatly lower the the energy needed to convert energy into matter and it is a lot less energy intense to make energy into the state of matter of gas than it is to make a solid or liquid - Nanotrasen uses this process to great effect on its research stations with long-term missions.


There, your lore is fixed.

**Comment by Iamgoofball** [comment]
basically

its a gift from the space wizard fed back when we were on good terms with them

**Comment by bgobandit** [comment]
how much power does it use up, it should use up a fuckton imo

**Comment by kevinz000** [comment]
@bgobandit not that much right now, it's intended to be scaled based on setting later whe nI learn enough TGUI to make them CONTROLLABLE! But it'll probably be first put on the map at a static power usage.

**Comment by Tokiko1** [comment]
What if you could pump infinite amounts of energy into this and get gas based on the amount of energy you put in?

**Comment by ExcessiveUseOfCobblestone** [comment]
turbine is now objectively the best engine!

**Comment by kevinz000** [comment]
@Tokiko1 someone make me a tgui interface/hold me at gunpoint to learn it and make one and sure :P

**Comment by kevinz000** [comment]
ok we're all good currently these are adminspawn, adding them to the map will come after this PR is reviewed/gotten over with, because i don't want to have to deal with a million lines of map conflicts and cry.

File: code/modules/atmospherics/machinery/other/miner.dm [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	return TRUE
+
+/obj/machinery/atmospherics/miner/proc/update_power()
+	if(!active)
+		active_power_usage = idle_power_usage
+	var/turf/T = get_turf(src)
+	var/datum/gas_mixture/G = T.return_air()
+	var/P = G.return_pressure()
+	switch(power_draw)
+		if(0)
```
Comment (lzimann) [inline code/modules/atmospherics/machinery/other/miner.dm]: use defines for this

File: code/modules/atmospherics/machinery/other/miner.dm [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	var/datum/gas_mixture/merger = new
+	merger.assert_gas(spawn_id)
+	merger.gases[spawn_id][MOLES] = (spawn_mol)
+	merger.temperature = spawn_temp
+	O.assume_air(merger)
+	SSair.add_to_active(O)
+
+/obj/machinery/atmospherics/miner/attack_ai(mob/living/silicon/user)
+	if(broken)
+		user << "[src] seems to be broken. Its debug interface outputs: [broken_message]"
```
Comment (lzimann) [inline code/modules/atmospherics/machinery/other/miner.dm]: Use `to_chat(src, message)`

**Comment by kevinz000** [comment]
fixed leoz

File: code/modules/atmospherics/machinery/other/miner.dm [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	var/power_draw_dynamic_kpa_coeff = 0.5
+	var/broken = FALSE
+	var/broken_message = "ERROR"
+	idle_power_usage = 150
+	active_power_usage = 2000
+
+/obj/machinery/atmospherics/miner/examine(mob/user)
+	..()
+	if(broken)
+		user << "Its debug output is printing \"[broken_message]\""
```
Comment (lzimann) [inline code/modules/atmospherics/machinery/other/miner.dm]: to_chat here too(don't know how i missed this in the first time)
